File: python/src/anysearch/providers/brave.py
"""Brave Search API — independent index with web + news search."""

from __future__ import annotations

from datetime import date

import httpx

from .._http import PreparedRequest
from ..types import SearchRequest, SearchResponse
from .base import BaseProvider, Capability


class BraveProvider(BaseProvider):
    name = "brave"
    aliases = ("brave_search",)
    env_keys = ("BRAVE_API_KEY", "BRAVE_SEARCH_API_KEY")
    default_base_url = "https://api.search.brave.com"
    capabilities = frozenset(
        {
            Capability.COUNTRY,
            Capability.LANGUAGE,
            Capability.DATE,
            Capability.SAFE_SEARCH,
            Capability.HIGHLIGHTS,
            Capability.NEWS,
        }
    )

    def prepare(self, req: SearchRequest) -> PreparedRequest:
        is_news = req.search_type == "news"
        max_count = 50 if is_news else 20
        params = {
            "q": req.query,
            "count": min(req.max_results, max_count),
        }
        if req.country:
            params["country"] = req.country.upper()
        if req.language:
            params["search_lang"] = req.language
        if req.safe_search:
            params["safesearch"] = req.safe_search
        if req.highlights:
            params["extra_snippets"] = "true"
        if req.start_published_date or req.end_published_date:
            start = req.start_published_date or "1970-01-01"
            end = req.end_published_date or date.today().isoformat()
            params["freshness"] = f"{start[:10]}to{end[:10]}"
        params.update(req.extra)

        path = "/res/v1/news/search" if is_news else "/res/v1/web/search"
        return PreparedRequest(
            method="GET",
            url=f"{self.base_url}{path}",
            headers={"X-Subscription-Token": self.api_key, "Accept": "application/json"},
            params=params,
        )

    def parse(self, response: httpx.Response, req: SearchRequest, elapsed_ms: float) -> SearchResponse:
        data = response.json()
        if req.search_type == "news":
            items = (data.get("results") or [])
        else:
            items = (data.get("web") or {}).get("results", []) or []
        results = []
        for item in items:
            extra_snippets = item.get("extra_snippets") or []
            results.append(
                self._result(
                    title=item.get("title"),
                    url=item.get("url"),
                    snippet=item.get("description"),
                    highlights=list(extra_snippets),
                    published_date=item.get("page_age") or item.get("age"),
                    raw=item,
                )
            )
        return self._response(req, results, data, elapsed_ms=elapsed_ms)

File: js/src/providers/brave.ts
import type { HttpResponse } from "../http.js";
import type { SearchRequest, SearchResponse } from "../types.js";
import { finalize, result, type ProviderContext, type ProviderSpec } from "./base.js";

export const brave: ProviderSpec = {
  name: "brave",
  aliases: ["brave_search"],
  envKeys: ["BRAVE_API_KEY", "BRAVE_SEARCH_API_KEY"],
  defaultBaseUrl: "https://api.search.brave.com",
  capabilities: new Set(["country", "language", "date", "safe_search", "highlights", "news"]),

  prepare(ctx: ProviderContext, req: SearchRequest) {
    const isNews = req.searchType === "news";
    const maxCount = isNews ? 50 : 20;
    const params: Record<string, string | number | boolean | undefined> = {
      q: req.query,
      count: Math.min(req.maxResults, maxCount),
    };
    if (req.country) params.country = req.country.toUpperCase();
    if (req.language) params.search_lang = req.language;
    if (req.safeSearch) params.safesearch = req.safeSearch;
    if (req.highlights) params.extra_snippets = "true";
    if (req.startPublishedDate || req.endPublishedDate) {
      const start = (req.startPublishedDate ?? "1970-01-01").slice(0, 10);
      const end = (req.endPublishedDate ?? new Date().toISOString()).slice(0, 10);
      params.freshness = `${start}to${end}`;
    }
    for (const [k, v] of Object.entries(req.extra)) params[k] = v as string;

    const path = isNews ? "/res/v1/news/search" : "/res/v1/web/search";
    return {
      method: "GET",
      url: `${ctx.baseUrl}${path}`,
      headers: { "X-Subscription-Token": ctx.apiKey ?? "", Accept: "application/json" },
      params,
    };
  },

  parse(_ctx, res: HttpResponse, req: SearchRequest, elapsedMs: number): SearchResponse {
    const data = res.data ?? {};
    const items = req.searchType === "news" ? data.results ?? [] : data.web?.results ?? [];
    const results = items.map((item: any) =>
      result({
        title: item.title,
        url: item.url,
        snippet: item.description,
        highlights: item.extra_snippets ?? [],
        publishedDate: item.page_age ?? item.age,
        raw: item,
      }),
    );
    return finalize(brave, req, results, data, { elapsedMs });
  },
};
